metricbeat/module/consul/agent/data.go

```diff
@@ -0,0 +1,133 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type valueHelper struct {
+	Value string
+	Unit  string
+}
+
+var (
+	allowedValues = map[string]valueHelper{
+		"consul.autopilot.healthy":         {Value: "autopilot.healthy"},
+		"consul.runtime.alloc_bytes":       {Value: "runtime.alloc", Unit: ".bytes"},
+		"consul.runtime.total_gc_pause_ns": {Value: "runtime.garbage_collector.pause.total", Unit: ".ns"},
+		"consul.runtime.gc_pause_ns":       {Value: "runtime.garbage_collector.pause.current", Unit: ".ns"},
+		"consul.runtime.total_gc_runs":     {Value: "runtime.garbage_collector.runs"},
+		"consul.runtime.num_goroutines":    {Value: "runtime.goroutines"},
+		"consul.runtime.heap_objects":      {Value: "runtime.heap_objects"},
+		"consul.runtime.sys_bytes":         {Value: "runtime.sys", Unit: ".bytes"},
+		"consul.runtime.malloc_count":      {Value: "runtime.malloc_count"},
+	}
+	allowedDetailedValues = map[string]valueHelper{
+		"consul.raft.apply":      {Value: "raft.apply"},
+		"consul.raft.commitTime": {Value: "raft.commit_time", Unit: ".ms"},
+	}
+)
+
+func eventMapping(content []byte) ([]common.MapStr, error) {
+	var agent agent
+
+	if err := json.Unmarshal(content, &agent); err != nil {
+		return nil, err
+	}
+
+	labels := map[string]common.MapStr{}
+
+	for _, gauge := range agent.Gauges {
+		metricApply(labels, gauge.consulMetric, gauge.Value)
+	}
+
+	for _, point := range agent.Points {
+		metricApply(labels, point.consulMetric, point.Value)
+	}
+
+	for _, counter := range agent.Counters {
+		metricApply(labels, counter.consulMetric, consulDetailedValue(counter))
+	}
+
+	for _, sample := range agent.Samples {
+		metricApply(labels, sample.consulMetric, consulDetailedValue(sample))
+	}
+
+	data := make([]common.MapStr, 0)
+	for _, v := range labels {
+		data = append(data, v)
+	}
+
+	return data, nil
+}
+
+func metricApply(labels map[string]common.MapStr, m consulMetric, v interface{}) {
+	prettyName := prettyName(m.Name)
+	if prettyName == nil {
+		//omitting unwanted metric
+		return
+	}
+
+	labelsCombination := ""
+	for k, v := range m.Labels {
+		labelsCombination = fmt.Sprintf("%s%s%s", labelsCombination, k, v)
+	}
```

Comment:
In other places we use [github.com/mitchellh/hashstructure](https://github.com/mitchellh/hashstructure) to get a hash from a map.

Reply:
I see no benefit of using a third party library here:
* A hash might have unwanted collisions while with this way a collision is totally wanted.
* No unnecessary third party libraries for a single line of code in the entire module
* In any case, it's not a hash of a map but a hash of 3 specific values. I'd need to create an anonymous struct to use this library

So I'll leave it like this. Thanks for the suggestion anyways :)

Reply:
> I see no benefit of using a third party library here:
> 
>     * A hash might have unwanted collisions while with this way a collision is totally wanted.
> 

A hash gives the same result for an equivalent input, and this is what we want. I think we can assume here the risk of collisions of a hash function. This one in particular also has options to reduce probability of collisions in specific data sets.

There is another alternative, use `common.MapStr` as [prometheus module does](https://github.com/elastic/beats/blob/master/metricbeat/module/prometheus/collector/data.go#L37). It obtains a json string with ordered keys, no chance of collision there, and the result is the same for equivalent inputs.

>     * No unnecessary third party libraries for a single line of code in the entire module
> 

Using tested third party libraries avoids problems like this: https://play.golang.org/p/MqYLTs-GNWf (I assume we want to put metrics with same labels in the same events).

Using `hashstructure` or `common.MapStr` you wouldn't be adding any dependency because we already use these methods in other places, even for similar purpouses.

>     * In any case, it's not a hash of a map but a hash of 3 specific values. I'd need to create an anonymous struct to use this library
> 

Umm, not sure to understand this comment, but the library supports (and I quote the first line of the readme) _creating a unique hash value for arbitrary values in Go_. With a small test you can see it.

> So I'll leave it like this. Thanks for the suggestion anyways :)

Please, if you keep your own implementation for this, encapsulate it in a function and test it.

Sorry if I am not understanding something, but current implementation doesn't look correct and I don't see the problem with using a library function for this :slightly_smiling_face:

Reply:
Please, try to be neat in your responses, we are trying to make this PR progress and not waste time trying to be right over the other.

Main problem of hashes is that thay actually might have really unwanted collisions, Jaime https://en.wikipedia.org/wiki/Hash_function.

So, please, let's solve this in a Zoom chat